### PR TITLE
feat: add create repo from git service

### PR DESCRIPTION
**Pull request by Extheoisah**
This pr connects the git service to the backend server. It allows a user to create a repo from the git service and returns the repo URL to the client. 
This pr also adds a new endpoint that allows admins to add challenges (courses) to the database. The repository URL field of the challenge is the starter code URL hosted on GitHub. This URL is needed to create a new repository for the user. The git service creates the user repo off of the starter code provided which in this case is mapped to Github repository.

### CODE REVIEW

File: src/app/routes/repo.rs [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+        }
+    };
+    if repo_url.is_empty() {
+        return Err(RepositoryError::BadRequest(
+            "Repository url is required".to_string(),
+        ));
+    }
+
+    let client = reqwest::Client::new();
+    let git_service_url = std::env::var("GIT_SERVICE_URL").expect("GIT_SERVICE_URL not set");
```
Comment (daviroo) [inline src/app/routes/repo.rs]:
```suggestion
       let git_service_url = std::env::var("GIT_SERVICE_URL").map_err(|_| {
        error!("GIT_SERVICE_URL environment variable not set");
        RepositoryError::InternalServerError("Server configuration error".to_string())
    })?;
```

The `expect` will cause the application to panic and crash in the event this env variable is not present which will effectively bring down the entire core server.  Instead we should handle this case gracefully.

File: src/app/routes/repo.rs [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+            )));
+        }
+    };
+
+    let repo_name = repo_url
+        .split("/")
+        .last()
+        .ok_or(RepositoryError::BadRequest(
+            "Error getting repository name from url".to_string(),
+        ))?;
```
Comment (daviroo) [inline src/app/routes/repo.rs]:
```suggestion
        .rsplit('/')
        .next()
        .and_then(|name| if name.is_empty() { None } else { Some(name) })
        .ok_or_else(|| {
            error!("Invalid repository URL format: {}", repo_url);
            RepositoryError::BadRequest("Invalid repository URL format".to_string())
        })?;
```

We should validate the format of the url before working on the string to ensure we get expected behaviour.

File: src/service/repository/users.rs [hunk 3]
```diff
@@ -60,27 +60,68 @@ impl User {
...
-            .map_err(|e| {
-                error!("Error getting user: {}", e);
-                FailedToGetUser(GetUserError(e))
-            })?;
 
-        user.ok_or_else(|| {
-            error!("User not found");
-            UserNotFound.into()
-        })
+        match (user_id, username, email, github_username) {
```
Comment (daviroo) [inline src/service/repository/users.rs:83]: The current `get_user` implementation has two issues:

1. It doesn't correctly handle cases where 2-3 parameters are provided.
2. The error message is misleading when multiple parameters are given.

**Review by daviroo — CHANGES_REQUESTED** [review]
Some suggestions for the repo route and the user repo.

File: src/app/routes/repo.rs (hunk 1 above)
Comment (Extheoisah) [reply src/app/routes/repo.rs]: Shouldn't it be the desired behaviour for the application to crash if any environment variables are missing? The application won't work so don't you think it makes sense to panic and throw the error?

Comment (Extheoisah) [reply src/app/routes/repo.rs]: How about adding a check in the entrypoint (main.rs) for the env variables needed to run the application. If any env variable is missing, we panic and don't start the server.
E.g.
```rust
async fn main() -> std::io::Result<()> {
    dotenv().ok();
    // check if any env's are not set
    if std::env::var("DATABASE_URL").is_err() {
        panic!("DATABASE_URL is not set");
    }
    if std::env::var("GIT_SERVICE_URL").is_err() {
        panic!("GIT_SERVICE_URL is not set");
    }
    env_logger::Builder::from_env(Env::default().default_filter_or("debug")).init();

    let pool = get_connection_pool();
    let manager_handle = WebSocketManagerHandle::new();

    HttpServer::new(move || {
        App::new()
            .app_data(web::Data::new(pool.clone()))
            .app_data(web::Data::new(manager_handle.clone()))
            .wrap(Logger::default())
            .service(
                web::scope("/api")
                    .wrap(AuthMiddleware)
                    .configure(routes::init),
            )
            .service(
                web::resource("/ws")
                    .wrap(AuthMiddleware)
                    .route(web::get().to(websocket_handler)),
            )
            .service(
                web::resource("/webhook").route(web::post().to(routes::webhook::handle_webhook)),
            )
    })
    .bind("127.0.0.1:4925")?
    .run()
    .await
}
```

File: src/service/repository/users.rs (hunk 3 above)
Comment (Extheoisah) [reply src/service/repository/users.rs:83]: addressed https://github.com/Extheoisah/hxckr-core/pull/19/commits/a110e98ad5bdbd5b945d61bb8b9206a37b54eb98

File: src/app/routes/repo.rs (hunk 2 above)
Comment (Extheoisah) [reply src/app/routes/repo.rs]: addressed https://github.com/Extheoisah/hxckr-core/pull/19/commits/3f24e614b071801e3ee2d11924ee8e6da32f700c

File: src/app/routes/repo.rs (hunk 1 above)
Comment (daviroo) [reply src/app/routes/repo.rs]: Yes that's better 👍

**Review by daviroo — APPROVED** [review]

**Review by daviroo — APPROVED** [review]

**Pull request merged**